web/packages/teleport/src/Welcome/NewCredentials/types.ts

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Attempt } from 'shared/hooks/useAttemptNext';
+
+import { Auth2faType, PrimaryAuthType } from 'shared/services';
+
+import { NewFlow, StepComponentProps } from 'design/StepSlider';
+
+import { ReactElement } from 'react';
+
+import { RecoveryCodes, ResetToken } from 'teleport/services/auth';
+
+export type UseTokenState = {
+  auth2faType: Auth2faType;
+  primaryAuthType: PrimaryAuthType;
+  isPasswordlessEnabled: boolean;
+  fetchAttempt: Attempt;
+  submitAttempt: Attempt;
+  clearSubmitAttempt: () => void;
+  onSubmit: (password: string, otpCode?: string, deviceName?: string) => void;
+  onSubmitWithWebauthn: (password?: string, deviceName?: string) => void;
+  resetToken: ResetToken;
+  recoveryCodes: RecoveryCodes;
+  redirect: () => void;
+  success: boolean;
+  finishedRegister: () => void;
+  privateKeyPolicyEnabled: boolean;
+};
+
+// duplicated from E
+export type QuestionnaireProps = {
+  full: boolean;
+  username: string;
+  onSubmit?: () => void;
+};
+
+export type NewCredentialsProps = UseTokenState & {
+  resetMode?: boolean;
+  isDashboard: boolean;
+
+  // support E questionnaire
+  displayOnboardingQuestionnaire?: boolean;
+  setDisplayOnboardingQuestionnaire?: (bool: boolean) => void;
+  Questionnaire?: (props: QuestionnaireProps) => ReactElement;
```

Comment:
A next-level improvement would be to introduce another layer of wrapping.

At the moment we got rid of having to duplicate the same component between OSS and E. However, `NewCredentials` in OSS contains some logic which is only relevant for E.

Instead of accepting a `Questionnaire`, `NewCredentials` could accept `RegisterSuccess`. The OSS version would then pass regular `RegisterSuccess`, while E would pass a special wrapper that renders either regular `RegisterSuccess` or `Questionnaire`.

This way `NewCredentials` in OSS wouldn't need to worry about `displayOnboardingQuestionnaire`, `setDisplayOnboardingQuestionnaire` & `Questionnaire`. There would also be no need to duplicate `QuestionnaireProps`. OSS would be simply oblivious to what happens in E.

Reply:
I'm adding this as a task on my parent/tracking issue!